File: packages/llm-test-harness/src/metrics/tool-called.ts
import type { MetricFn, MetricScore } from '../types.js'

interface ContentBlock {
  type?: string
  name?: string
  input?: unknown
}

/**
 * Asserts that the LLM response included a tool call with the given name.
 *
 * Pass the response content as JSON:
 * @example
 * const text = JSON.stringify(response.content)
 * harness.evaluate(text, [Metrics.toolCalled('search')])
 *
 * Optionally verify the tool input contains a substring:
 * @example
 * Metrics.toolCalled('search', { inputContains: 'Paris' })
 */
export class ToolCalled implements MetricFn {
  readonly name: string
  private toolName: string
  private inputContains?: string

  constructor(toolName: string, options?: { inputContains?: string }) {
    this.toolName = toolName
    this.inputContains = options?.inputContains
    this.name = `ToolCalled(${toolName})`
  }

  evaluate(text: string): MetricScore {
    let blocks: ContentBlock[]
    try {
      const parsed = JSON.parse(text)
      blocks = Array.isArray(parsed) ? parsed : [parsed]
    } catch {
      return {
        name: this.name,
        pass: false,
        score: 0,
        reason: 'Input is not valid JSON',
      }
    }

    const toolBlock = blocks.find(
      (b) => b.type === 'tool_use' && b.name === this.toolName,
    )

    if (!toolBlock) {
      const found = blocks.filter((b) => b.type === 'tool_use').map((b) => b.name)
      return {
        name: this.name,
        pass: false,
        score: 0,
        reason: found.length
          ? `Tool "${this.toolName}" not called. Calls found: ${found.join(', ')}`
          : `No tool calls found in response`,
      }
    }

    if (this.inputContains) {
      const inputStr = JSON.stringify(toolBlock.input ?? '')
      if (!inputStr.includes(this.inputContains)) {
        return {
          name: this.name,
          pass: false,
          score: 0.5,
          reason: `Tool "${this.toolName}" was called but input does not contain "${this.inputContains}"`,
        }
      }
    }

    return { name: this.name, pass: true, score: 1 }
  }
}

File: packages/llm-test-harness-python/llm_test_harness/metrics/tool_called.py
from __future__ import annotations

import json
from typing import Any, Optional

from ..types import MetricFn, MetricScore


class ToolCalled(MetricFn):
    """Assert the LLM called a specific tool.

    Pass ``JSON.stringify(response.content)`` (or ``json.dumps(...)`` in Python)
    as the evaluated text.

    Example::

        text = json.dumps([b.__dict__ for b in response.content])
        harness.evaluate(text, [Metrics.tool_called("search")])
        harness.evaluate(text, [Metrics.tool_called("search", input_contains="Paris")])
    """

    def __init__(self, tool_name: str, *, input_contains: Optional[str] = None) -> None:
        self._tool_name = tool_name
        self._input_contains = input_contains

    @property
    def name(self) -> str:
        return f"ToolCalled({self._tool_name})"

    def evaluate(self, text: str) -> MetricScore:
        try:
            parsed: Any = json.loads(text)
            blocks: list[Any] = parsed if isinstance(parsed, list) else [parsed]
        except json.JSONDecodeError:
            return MetricScore(name=self.name, passed=False, score=0.0, reason="Input is not valid JSON")

        tool_block = next(
            (b for b in blocks if isinstance(b, dict) and b.get("type") == "tool_use" and b.get("name") == self._tool_name),
            None,
        )

        if tool_block is None:
            found = [b.get("name") for b in blocks if isinstance(b, dict) and b.get("type") == "tool_use"]
            reason = (
                f'Tool "{self._tool_name}" not called. Calls found: {", ".join(str(n) for n in found)}'
                if found
                else "No tool calls found in response"
            )
            return MetricScore(name=self.name, passed=False, score=0.0, reason=reason)

        if self._input_contains:
            input_str = json.dumps(tool_block.get("input", ""))
            if self._input_contains not in input_str:
                return MetricScore(
                    name=self.name,
                    passed=False,
                    score=0.5,
                    reason=f'Tool "{self._tool_name}" was called but input does not contain "{self._input_contains}"',
                )

        return MetricScore(name=self.name, passed=True, score=1.0)
